File: api/src/LingoCross.Application/Games/Dtos/GameDtos.cs
using LingoCross.Domain.Enums;

namespace LingoCross.Application.Games.Dtos;

/// <summary>Bir derse ait oyunun özet bilgisi (oluşturulduktan/yayımlandıktan sonra döndürülür).</summary>
public record GameDto(
    Guid Id,
    Guid LessonId,
    GameType Type,
    string Title,
    bool IsPublished,
    DateTime? PublishedAt,
    DateTime CreatedAt,
    DateTime UpdatedAt);

/// <summary>
/// Öğrenciye atanmış (yayımlanmış) bir oyunun özeti. Öğrenci panelinde "atanan bulmacalar"
/// listesini besler: oyun + ait olduğu ders + öğretmen adı + oynanabilir kelime sayısı.
/// Tamamlanma alanları öğrencinin bu oyunu zaten bitirip bitirmediğini taşır: tamamlanmışsa
/// <see cref="IsCompleted"/> true ve <see cref="ResultId"/>/<see cref="Score"/>/<see cref="CompletedAt"/>
/// dolar; aksi halde false/null. Tamamlanmış bir oyun tekrar oynanamaz (StartSession 409 döner).
/// </summary>
public record AssignedGameDto(
    Guid Id,
    Guid LessonId,
    string LessonTitle,
    GameType Type,
    string Title,
    int WordCount,
    string TeacherName,
    DateTime? PublishedAt,
    bool IsCompleted,
    Guid? ResultId,
    int? Score,
    DateTime? CompletedAt);

/// <summary>
/// Öğretmenin "Bulmacalarım" görünümü için tüm derslerindeki bir bulmacanın özeti. Yayımlanmış bir
/// bulmaca, öğretmenin tüm Active eşleşmeli öğrencilerine atanmış sayılır (<see cref="AssignedStudentCount"/>);
/// <see cref="SolveCount"/> bu bulmacaya ait tamamlanmış sonuç (game_results) sayısıdır.
/// </summary>
public record TeacherPuzzleDto(
    Guid Id,
    Guid LessonId,
    string LessonTitle,
    GameType Type,
    bool IsPublished,
    DateTime CreatedAt,
    int AssignedStudentCount,
    int SolveCount);

/// <summary>
/// Öğretmenin bir derste oyun oluşturma/yayımlama isteği. <see cref="ClassIds"/> verilirse (F4.3)
/// oyun, oluşturma ile aynı işlemde bu sınıflara atanır (set semantiği). null → atama dokunulmaz.
/// </summary>
public record CreateGameRequest(
    GameType Type,
    string? Title,
    IReadOnlyList<Guid>? ClassIds = null);

/// <summary>Bir oyunun atandığı sınıf kimlikleri (F4.3, set semantiği).</summary>
public record GameAssignmentsDto(
    IReadOnlyList<Guid> ClassIds);

/// <summary>Bir oyunu sınıflara atama isteği (gönderilen liste nihaidir; set semantiği).</summary>
public record SetGameAssignmentsRequest(
    IReadOnlyList<Guid> ClassIds);

/// <summary>
/// Öğretmenin bir oyun (bulmaca) için, kaydetmeden ÖNCE örnek içerik önizleme isteği. Yalnız tür
/// taşınır; başlık/atama yok (önizleme kalıcı değildir).
/// </summary>
public record PreviewGameRequest(
    GameType Type);

/// <summary>
/// Öğretmen önizleme yanıtı: oturum OLMADAN üretilmiş örnek oyun içeriği (<see cref="StartGameSessionResponse"/>
/// deseni, ama session yok). İçerik <see cref="Type"/>'a göre tür-duyarlıdır:
/// <list type="bullet">
///   <item><see cref="GameType.WordMatching"/> → <see cref="WordMatching"/> doludur (diğeri null).</item>
///   <item><see cref="GameType.Crossword"/> → <see cref="Crossword"/> doludur (diğeri null).</item>
/// </list>
/// Hiçbir Game/GameSession kaydı oluşturulmaz; içerik StartSession ile aynı yeterlilik kurallarına tabidir.
/// </summary>
public record GamePreviewResponse(
    GameType Type,
    WordMatchingContent? WordMatching,
    CrosswordContent? Crossword,
    ScrambledContent? Scrambled);

/// <summary>Bir oyun oturumunun durumu.</summary>
public record GameSessionDto(
    Guid Id,
    Guid GameId,
    Guid StudentId,
    GameSessionStatus Status,
    DateTime StartedAt,
    DateTime? CompletedAt);

/// <summary>
/// Oturum başlatma yanıtı: oturum bilgisi + üretilmiş oyun içeriği. İçerik <see cref="Type"/>'a göre
/// tür-duyarlıdır; istemci türü okuyup yalnızca ilgili alanı kullanır:
/// <list type="bullet">
///   <item><see cref="GameType.WordMatching"/> → <see cref="WordMatching"/> doludur (diğeri null).</item>
///   <item><see cref="GameType.Crossword"/> → <see cref="Crossword"/> doludur (diğeri null).</item>
/// </list>
/// Geriye dönük uyumluluk: WordMatching oturumlarında <see cref="WordMatching"/> önceki
/// <c>Content</c> alanının aynısıdır (aynı şekil: pairs + distractors).
/// </summary>
public record StartGameSessionResponse(
    GameSessionDto Session,
    GameType Type,
    WordMatchingContent? WordMatching,
    CrosswordContent? Crossword,
    ScrambledContent? Scrambled);

/// <summary>
/// Kelime eşleştirme oyun içeriği. <see cref="Pairs"/> doğru eşleşmeleri taşır; her çiftin
/// birincil Türkçe karşılığı <see cref="MatchingPair.CorrectTranslation"/>'dadır.
/// <see cref="Distractors"/>, başka kelimelerin çevirilerinden seçilmiş ve doğru cevap olmayan
/// ek Türkçe karşılıklardır (sağ sütunu zorlaştırmak için). Tüm sağ-sütun adayları
/// (doğru çeviriler + çeldiriciler) benzersizdir.
/// </summary>
public record WordMatchingContent(
    IReadOnlyList<MatchingPair> Pairs,
    IReadOnlyList<string> Distractors);

/// <summary>Tek bir eşleştirme çifti: kaynak terim ve onun doğru birincil çevirisi.</summary>
public record MatchingPair(
    Guid WordId,
    string Term,
    string CorrectTranslation);

/// <summary>
/// Crossword (bulmaca) oyun içeriği. Izgara <see cref="Rows"/>×<see cref="Cols"/> hücredir
/// (0-tabanlı, satır-major). Her giriş bir kelimeyi temsil eder:
/// <para>
/// - <see cref="CrosswordEntry.Answer"/> doğru cevaptır: yalnız A–Z BÜYÜK harfler (İngilizce terimden
///   normalize). MVP'de cevap istemciye gönderilir; doğrulama istemcide yapılır.<br/>
/// - <see cref="CrosswordEntry.Clue"/> ipucudur: terimin birincil Türkçe karşılığı.<br/>
/// - <see cref="CrosswordEntry.Row"/>/<see cref="CrosswordEntry.Col"/> kelimenin BAŞLANGIÇ hücresidir
///   (0-tabanlı). across (yatay) ise harfler sütun artarak, down (dikey) ise satır artarak yerleşir.<br/>
/// - <see cref="CrosswordEntry.Direction"/>: 0 = across (yatay), 1 = down (dikey).<br/>
/// - <see cref="CrosswordEntry.Number"/> klasik bulmaca numarasıdır; başlangıç hücreleri satır-major
///   sırada numaralanır, aynı hücreden hem across hem down başlıyorsa ikisi aynı numarayı paylaşır.
/// </para>
/// İstemci ızgarayı şöyle kurabilir: <see cref="Rows"/>×<see cref="Cols"/> boş ızgara aç; her giriş
/// için başlangıç hücresinden yönüne göre <see cref="CrosswordEntry.Length"/> hücreyi işaretle.
/// Kesişen hücrelerde harfler tutarlıdır (üretim bunu garanti eder).
/// </summary>
public record CrosswordContent(
    int Rows,
    int Cols,
    IReadOnlyList<CrosswordEntry> Entries);

/// <summary>Bulmacadaki tek bir kelime girişi. Bkz. <see cref="CrosswordContent"/>.</summary>
public record CrosswordEntry(
    int Number,
    string Answer,
    string Clue,
    int Row,
    int Col,
    CrosswordDirection Direction,
    int Length);

/// <summary>
/// Scrambled (harf karıştırma) oyun içeriği. Her <see cref="ScrambledItem"/> bir kelimeyi temsil eder:
/// kaynak terimin harfleri karıştırılmış olarak verilir, öğrenci doğru sıraya dizer.
/// </summary>
public record ScrambledContent(
    IReadOnlyList<ScrambledItem> Items);

/// <summary>
/// Tek bir scrambled kelime. <see cref="Answer"/> doğru cevaptır (kaynak terim, olduğu gibi),
/// <see cref="ScrambledLetters"/> aynı harflerin karıştırılmış hâlidir (çoklu-küme olarak Answer ile aynı),
/// <see cref="Clue"/> ipucudur (terimin birincil Türkçe karşılığı). MVP'de cevap istemciye gönderilir;
/// doğrulama istemcide yapılır.
/// </summary>
public record ScrambledItem(
    Guid WordId,
    string Answer,
    string ScrambledLetters,
    string Clue);

/// <summary>Bulmaca kelime yönü. Sayısal değerler istemciyle paylaşılır.</summary>
public enum CrosswordDirection
{
    /// <summary>Yatay (soldan sağa); harfler sütun artarak yerleşir.</summary>
    Across = 0,

    /// <summary>Dikey (yukarıdan aşağı); harfler satır artarak yerleşir.</summary>
    Down = 1,
}

File: mobile/lingo_cross_app/lib/features/games/data/dtos/game_dtos.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../domain/game_type.dart';

part 'game_dtos.freezed.dart';
part 'game_dtos.g.dart';

/// API'deki int `GameType` değerini ([GameType]) enum'una çevirir.
class GameTypeConverter implements JsonConverter<GameType, int> {
  const GameTypeConverter();

  @override
  GameType fromJson(int json) => GameType.fromValue(json);

  @override
  int toJson(GameType object) => object.value;
}

/// API'deki int `GameSessionStatus` değerini ([GameSessionStatus]) enum'una çevirir.
class GameSessionStatusConverter
    implements JsonConverter<GameSessionStatus, int> {
  const GameSessionStatusConverter();

  @override
  GameSessionStatus fromJson(int json) => GameSessionStatus.fromValue(json);

  @override
  int toJson(GameSessionStatus object) => object.value;
}

/// Bir derse ait oyunun özeti (GameDto) — API ile birebir.
@freezed
class GameDto with _$GameDto {
  const factory GameDto({
    required String id,
    required String lessonId,
    @GameTypeConverter() required GameType type,
    required String title,
    required bool isPublished,
    DateTime? publishedAt,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _GameDto;

  factory GameDto.fromJson(Map<String, dynamic> json) =>
      _$GameDtoFromJson(json);
}

/// Öğrenciye atanmış (yayımlanmış) bir oyunun özeti (AssignedGameDto) — API ile
/// birebir. Öğrenci panelindeki "atanan bulmacalar" listesini besler.
///
/// [isCompleted] true ise bulmaca tamamlanmıştır (tekrar oynanamaz; yalnız
/// istatistik). Tamamlanmış oyunlarda [resultId]/[score]/[completedAt] doludur
/// ve dokununca `GET /results`'ten istatistik ekranına gidilir (replay değil).
@freezed
class AssignedGameDto with _$AssignedGameDto {
  const factory AssignedGameDto({
    required String id,
    required String lessonId,
    required String lessonTitle,
    @GameTypeConverter() required GameType type,
    required String title,
    required int wordCount,
    required String teacherName,
    DateTime? publishedAt,
    @Default(false) bool isCompleted,
    String? resultId,
    int? score,
    DateTime? completedAt,
  }) = _AssignedGameDto;

  factory AssignedGameDto.fromJson(Map<String, dynamic> json) =>
      _$AssignedGameDtoFromJson(json);
}

/// Öğretmenin "Bulmacalarım" görünümü için tüm derslerindeki bir bulmacanın
/// özeti (TeacherPuzzleDto) — API ile birebir.
///
/// [assignedStudentCount] yayımlı bulmacanın atandığı Active öğrenci sayısı
/// (bireysel model; sınıf/grup yok). [solveCount] bu bulmacaya ait tamamlanmış
/// sonuç (game_results) sayısıdır.
@freezed
class TeacherPuzzleDto with _$TeacherPuzzleDto {
  const factory TeacherPuzzleDto({
    required String id,
    required String lessonId,
    required String lessonTitle,
    @GameTypeConverter() required GameType type,
    required bool isPublished,
    required DateTime createdAt,
    required int assignedStudentCount,
    required int solveCount,
  }) = _TeacherPuzzleDto;

  factory TeacherPuzzleDto.fromJson(Map<String, dynamic> json) =>
      _$TeacherPuzzleDtoFromJson(json);
}

/// Öğretmenin bir derste oyun oluşturma/yayımlama isteği (CreateGameRequest).
///
/// [classIds] opsiyonel: F4.3 ile oyun oluştururken hedef sınıflar atanır
/// (`POST /lessons/{id}/games` body'sine eklenir). Boş/null ise mevcut davranış.
@freezed
class CreateGameRequest with _$CreateGameRequest {
  const factory CreateGameRequest({
    @GameTypeConverter() required GameType type,
    String? title,
    List<String>? classIds,
  }) = _CreateGameRequest;

  factory CreateGameRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateGameRequestFromJson(json);
}

/// Bir oyun oturumunun durumu (GameSessionDto) — API ile birebir.
@freezed
class GameSessionDto with _$GameSessionDto {
  const factory GameSessionDto({
    required String id,
    required String gameId,
    required String studentId,
    @GameSessionStatusConverter() required GameSessionStatus status,
    required DateTime startedAt,
    DateTime? completedAt,
  }) = _GameSessionDto;

  factory GameSessionDto.fromJson(Map<String, dynamic> json) =>
      _$GameSessionDtoFromJson(json);
}

/// Tek bir eşleştirme çifti: kaynak terim + doğru birincil çeviri (MatchingPair).
@freezed
class MatchingPair with _$MatchingPair {
  const factory MatchingPair({
    required String wordId,
    required String term,
    required String correctTranslation,
  }) = _MatchingPair;

  factory MatchingPair.fromJson(Map<String, dynamic> json) =>
      _$MatchingPairFromJson(json);
}

/// Kelime eşleştirme oyun içeriği (WordMatchingContent).
///
/// [pairs] doğru eşleşmeleri taşır; [distractors] hiçbir terimle eşleşmeyen
/// ek (çeldirici) Türkçe karşılıklardır. İstemci sağ sütunu
/// (doğru çeviriler + çeldiriciler) karıştırarak kurar.
@freezed
class WordMatchingContent with _$WordMatchingContent {
  const factory WordMatchingContent({
    required List<MatchingPair> pairs,
    required List<String> distractors,
  }) = _WordMatchingContent;

  factory WordMatchingContent.fromJson(Map<String, dynamic> json) =>
      _$WordMatchingContentFromJson(json);
}

/// Scrambled (karışık harfler) oyununda tek bir kelime (ScrambledItem) — API ile
/// birebir.
///
/// [answer] kaynak (doğru) terim; öğrenci bunu yeniden dizer. [scrambledLetters]
/// karışık harflerin tek string hâli (ör. "ppela"); istemci karaktere bölerek
/// havuzu kurar. [clue] hedef dildeki çeviri (ipucu olarak gösterilir).
@freezed
class ScrambledItem with _$ScrambledItem {
  const factory ScrambledItem({
    required String wordId,
    required String answer,
    required String scrambledLetters,
    required String clue,
  }) = _ScrambledItem;

  factory ScrambledItem.fromJson(Map<String, dynamic> json) =>
      _$ScrambledItemFromJson(json);
}

/// Scrambled (karışık harfler) oyun içeriği (ScrambledContent) — API ile birebir.
///
/// [items] her biri bir kelimeyi temsil eder (karışık harfler + çeviri ipucu).
@freezed
class ScrambledContent with _$ScrambledContent {
  const factory ScrambledContent({
    required List<ScrambledItem> items,
  }) = _ScrambledContent;

  factory ScrambledContent.fromJson(Map<String, dynamic> json) =>
      _$ScrambledContentFromJson(json);
}

/// Çoktan seçmeli soruda tek bir şık (QuestionChoice) — API ile birebir.
///
/// [optionId] şıkkın kalıcı kimliği (doğru cevap eşleşmesi bununla yapılır).
/// [label] görünür harf etiketi ("A".."E"). [text] şık metni.
@freezed
class QuestionChoice with _$QuestionChoice {
  const factory QuestionChoice({
    required String optionId,
    required String label,
    required String text,
  }) = _QuestionChoice;

  factory QuestionChoice.fromJson(Map<String, dynamic> json) =>
      _$QuestionChoiceFromJson(json);
}

/// Çoktan seçmeli tek bir soru (QuestionItem) — API ile birebir.
///
/// [stem] soru kökü. [choices] A–E sırasında şıklar. [correctOptionId] doğru
/// şıkkın [QuestionChoice.optionId]'si (skorlama İSTEMCİDE bununla yapılır).
/// [explanation] opsiyonel açıklama (oyun sırasında gösterilmez; ileride
/// kullanılabilir).
@freezed
class QuestionItem with _$QuestionItem {
  const factory QuestionItem({
    required String questionId,
    required String stem,
    required List<QuestionChoice> choices,
    required String correctOptionId,
    String? explanation,
  }) = _QuestionItem;

  factory QuestionItem.fromJson(Map<String, dynamic> json) =>
      _$QuestionItemFromJson(json);
}

/// Çıkmış Sorular (çoktan seçmeli) oyun içeriği (QuestionSetContent) — API ile
/// birebir.
///
/// [questions] 10 soruluk set; her biri A–E şıklı tek seçimdir. Doğru cevap
/// içerikle gelir (skorlama istemcide); tekrar çözme backend'de tek sefer.
@freezed
class QuestionSetContent with _$QuestionSetContent {
  const factory QuestionSetContent({
    required List<QuestionItem> questions,
  }) = _QuestionSetContent;

  factory QuestionSetContent.fromJson(Map<String, dynamic> json) =>
      _$QuestionSetContentFromJson(json);
}

/// Öğretmenin atayabileceği bir soru konusu (QuestionTopicDto) — API ile birebir.
///
/// `GET /api/question-topics` döner. [questionCount] konudaki soru adedi (rozet).
@freezed
class QuestionTopicDto with _$QuestionTopicDto {
  const factory QuestionTopicDto({
    required String id,
    required String title,
    String? description,
    required int questionCount,
  }) = _QuestionTopicDto;

  factory QuestionTopicDto.fromJson(Map<String, dynamic> json) =>
      _$QuestionTopicDtoFromJson(json);
}

/// Bir oyunun/konunun sınıf atamalarının özeti (GameAssignmentsDto) — API ile
/// birebir.
///
/// `POST/GET /api/question-topics/{topicId}/assignments` yanıtı. [classIds]
/// konunun atandığı sınıf kimlikleridir.
@freezed
class GameAssignmentsDto with _$GameAssignmentsDto {
  const factory GameAssignmentsDto({
    @Default(<String>[]) List<String> classIds,
  }) = _GameAssignmentsDto;

  factory GameAssignmentsDto.fromJson(Map<String, dynamic> json) =>
      _$GameAssignmentsDtoFromJson(json);
}

/// API'deki int `CrosswordDirection` değerini ([CrosswordDirection]) enum'una çevirir.
class CrosswordDirectionConverter
    implements JsonConverter<CrosswordDirection, int> {
  const CrosswordDirectionConverter();

  @override
  CrosswordDirection fromJson(int json) => CrosswordDirection.fromValue(json);

  @override
  int toJson(CrosswordDirection object) => object.value;
}

/// Bulmacadaki tek bir kelime girişi (CrosswordEntry) — API ile birebir.
///
/// [answer] yalnız A–Z BÜYÜK harfler (doğrulama istemcide). [clue] Türkçe ipucu.
/// [row]/[col] başlangıç hücresi (0-tabanlı). across → harfler sütun artarak,
/// down → satır artarak yerleşir. [number] klasik bulmaca numarasıdır.
@freezed
class CrosswordEntry with _$CrosswordEntry {
  const factory CrosswordEntry({
    required int number,
    required String answer,
    required String clue,
    required int row,
    required int col,
    @CrosswordDirectionConverter() required CrosswordDirection direction,
    required int length,
  }) = _CrosswordEntry;

  factory CrosswordEntry.fromJson(Map<String, dynamic> json) =>
      _$CrosswordEntryFromJson(json);
}

/// Crossword (bulmaca) oyun içeriği (CrosswordContent) — API ile birebir.
///
/// Izgara [rows]×[cols] hücredir (0-tabanlı, satır-major). [entries] her biri bir
/// kelimeyi temsil eder. Kesişen hücrelerde harfler tutarlıdır (üretim garanti eder).
@freezed
class CrosswordContent with _$CrosswordContent {
  const factory CrosswordContent({
    required int rows,
    required int cols,
    required List<CrosswordEntry> entries,
  }) = _CrosswordContent;

  factory CrosswordContent.fromJson(Map<String, dynamic> json) =>
      _$CrosswordContentFromJson(json);
}

/// Oyun önizleme yanıtı (GamePreviewResponse): kaydetmeden önce üretilecek
/// örnek içerik. `POST /lessons/{lessonId}/games/preview` döner. Kalıcı değildir;
/// [type] WordMatching ise [wordMatching] doludur, Crossword ise [crossword].
/// İçerik şekilleri [StartGameSessionResponse] ile aynı DTO'ları kullanır.
@freezed
class GamePreviewResponse with _$GamePreviewResponse {
  const factory GamePreviewResponse({
    @GameTypeConverter() required GameType type,
    WordMatchingContent? wordMatching,
    CrosswordContent? crossword,
    ScrambledContent? scrambled,
    QuestionSetContent? questionSet,
  }) = _GamePreviewResponse;

  factory GamePreviewResponse.fromJson(Map<String, dynamic> json) =>
      _$GamePreviewResponseFromJson(json);
}

/// Oturum başlatma yanıtı (StartGameSessionResponse): oturum + tür-duyarlı içerik.
///
/// Yeni şekil: `{ session, type, wordMatching?, crossword? }`. [type] WordMatching
/// ise [wordMatching] doludur (eski `content` ile aynı şekil); Crossword ise
/// [crossword] doludur. İstemci [type]'a göre dallanır.
@freezed
class StartGameSessionResponse with _$StartGameSessionResponse {
  const factory StartGameSessionResponse({
    required GameSessionDto session,
    @GameTypeConverter() required GameType type,
    WordMatchingContent? wordMatching,
    CrosswordContent? crossword,
    ScrambledContent? scrambled,
    QuestionSetContent? questionSet,
  }) = _StartGameSessionResponse;

  factory StartGameSessionResponse.fromJson(Map<String, dynamic> json) =>
      _$StartGameSessionResponseFromJson(json);
}
